src/manage_sql_agents.c

```diff
@@ -0,0 +1,673 @@
+        insert_hostname,
+        agent->authorized,
+        agent->min_interval,
+        agent->heartbeat_interval,
+        insert_connection_status,
+        agent->last_update_agent_control,
+        insert_schedule,
+        agent->owner,
+        agent->modification_time,
+        agent->scanner,
+        insert_agent_id);
+
+  g_free (insert_hostname);
+  g_free (insert_connection_status);
+  g_free (insert_schedule);
+  g_free (insert_agent_id);
+}
+
+/**
+ * @brief Append an agent's data as a row to a COPY buffer.
+ *
+ * @param buffer COPY buffer for agents.
+ * @param agent  Agent whose data will be appended.
+ */
+static void
+append_agent_row_to_buffer (db_copy_buffer_t *buffer, agent_data_t agent)
+{
+  if (!agent->uuid)
+    {
+      agent->uuid = gvm_uuid_make ();
+      if (agent->uuid == NULL)
+        return;
+    }
+   gchar *escaped_hostname =  sql_copy_escape (agent->hostname);
+   gchar *escaped_connection_status =  sql_copy_escape (agent->connection_status);
+   gchar *escaped_schedule =  sql_copy_escape (agent->schedule);
+   gchar *escaped_comment =  sql_copy_escape ("");
+
+  db_copy_buffer_append_printf (
+    buffer,
+    "%s\t%s\t%s\t%s\t%d\t%d\t%d\t%s\t%ld\t%s\t%u\t%s\t%ld\t%ld\t%llu\n",
+    agent->uuid,
+    agent->name,
+    agent->agent_id,
+    escaped_hostname,
+    agent->authorized,
+    agent->min_interval,
+    agent->heartbeat_interval,
+    escaped_connection_status,
+    agent->last_update_agent_control,
+    escaped_schedule,
+    agent->owner,
+    escaped_comment,
+    time (NULL),     // creation time
+    agent->modification_time,
+    agent->scanner
+  );
+
+  g_free (escaped_hostname);
+  g_free (escaped_connection_status);
+  g_free (escaped_schedule);
+  g_free (escaped_comment);
+}
+
+/**
+ * @brief Append all IPs of an agent to a COPY buffer.
+ *
+ * @param buffer   COPY buffer for agent IPs.
+ * @param agent_id ID of the agent.
+ * @param ip_list  List of IP addresses to append.
+ */
+static void
+append_ip_rows_to_buffer (db_copy_buffer_t *buffer,
+                          const gchar *agent_id,
+                          agent_ip_data_list_t ip_list)
+{
+  if (!ip_list) return;
+
+  gchar *escaped_agent_id =  sql_copy_escape (agent_id);
+
+  for (int j = 0; j < ip_list->count; ++j)
+    {
+      agent_ip_data_t ip = ip_list->items[j];
+      gchar *escaped_ip_address =  sql_copy_escape (ip->ip_address);
+
+      db_copy_buffer_append_printf (
+        buffer,
+        "%s\t%s\n",
+        escaped_agent_id,
+        escaped_ip_address);
+
+      g_free (escaped_ip_address);
+    }
+
+  g_free (escaped_agent_id);
+}
+
+/**
+ * @brief Add an IP data entry to an agent_ip_data_list.
+ *
+ * @param list     Target IP list.
+ * @param ip_data  IP data to append.
+ */
+static void
+agent_ip_data_list_add (agent_ip_data_list_t list, agent_ip_data_t ip_data)
+{
+  g_return_if_fail (list);
+
+  list->items = g_realloc (list->items, (list->count + 1) * sizeof (agent_ip_data_t));
```

Comment:
I'm not sure if this already becomes relevant here, but if there are lots of these lists and they can get fairly long then it may be a good idea to avoid reallocations by keeping track of the allocated size and increasing it multiplicatively (for example doubling the size) if needed or switching to GPtrArray which handles this.